SEP-24: Un-deprecate `postMessage` callbacks and add security section (#1081)
stellar/stellar-protocol: ecosystem/sep-0024.md

`on_change_callback` | string | (optional) [`postMessage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) or a URL that the anchor should `POST` a JSON message to when the `status` or `kyc_verified` properties change.

The URL supplied by both callback parameters should receive the full transaction object.

**`callback` details**

If the wallet wants to be notified that the user has completed the anchor's interactive flow (either success or failure), it can add this parameter to the URL. If the user abandons the process, the anchor does not need to report anything to the wallet. If the `callback` value is a URL, the anchor must `POST` to it with a JSON message as the body.

**`postMessage` details**

If provided, the anchor must post a JSON message (either as a JSON-serialized string or a plain javascript object) to `window.opener` via the Javascript [`Window.postMessage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) method. If `window.opener` is undefined, the message must be posted to `window.parent` instead.

> Because `callback` is used to notify the wallet that the interactive flow is complete, it is common for anchors to make the `postMessage` callback as a result of the user clicking a "Close" button or similar UI element. The wallet can then close the window displaying the anchor's interactive flow using `[Window.close()`](https://developer.mozilla.org/en-US/docs/Web/API/Window/close). This ensures the wallet application receives the transaction's information and the user has a smooth experience.

Reviewer: 💡 Could you add a link to this section to the security concerns section? I know these sections are back to back, but sometimes we read docs with narrow vision and I think it is worth it.

Author: Sure thing 👍

Author: Actually, we cannot link to the security section because its just bolded text, not a header. I can mention there are security concerns though and to see the section below.

Reviewer: Ah okay, no need to go out of our way for this I think.

> **Interactive Flow Security Concerns**

Reviewer: 💡 Suggested title name to focus on the security concerns and impacted functionality. I.e. If I'm using postMessage this might trigger me to actually read the security concerns section.
```suggestion
**Security Concerns (`noopener`, `postMessage`)**
```

> Because wallet applications display content rendered by a third party anchor service, it is recommended to use the [`noopener`](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/noopener) link type or feature for [`Window.open()`](https://developer.mozilla.org/en-US/docs/Web/API/Window/open#window_functionality_features).

Reviewer: 💡 
```suggestion
Wallet applications display content rendered by a third party anchor service. It is recommended to use the [`rel=noopener`](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/noopener) attribute on links or the `noopener` feature for [`Window.open()`](https://developer.mozilla.org/en-US/docs/Web/API/Window/open#window_functionality_features).
```

> Note that using `noopener` prohibits the use of `postMessage` callbacks. If `postMessage` callbacks are required for your implementation, it recommended to only open URLs from known and trusted third party anchors.

Reviewer: 💡 
```suggestion
Note that using `noopener` prohibits the use of `postMessage` callbacks. If `postMessage` callbacks are required for your implementation, it recommended to only open URLs from anchors that you, the wallet developer, trust.
```
